edge/pkg/devicetwin/dtmanager/dmiserver.go

```diff
@@ -0,0 +1,427 @@
+	protocol, err := dtcommon.GetProtocolNameOfDevice(device)
+	if err != nil {
+		return err
+	}
+	sockPath := GetSockPath(protocol)
+	if sockPath == "" {
+		return fmt.Errorf("cannot get sockPath of protocol %s", protocol)
+	}
+
+	dc, ctx, conn, cancelFunc, err := dmiclient.GenerateDMIClient(sockPath)
+	if err != nil {
+		return err
+	}
+
+	defer conn.Close()
+	defer cancelFunc()
+
+	rdr, err := dmiclient.RemoveDeviceRequest(device.Name)
+	if err != nil {
+		return fmt.Errorf("fail to generate RemoveDeviceRequest for device %s with err: %v", device.Name, err)
+	}
+	_, err = dc.RemoveDevice(ctx, rdr)
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+func GrpcUpdateDevice(device *v1alpha2.Device) error {
+	protocol, err := dtcommon.GetProtocolNameOfDevice(device)
+	if err != nil {
+		return err
+	}
+	sockPath := GetSockPath(protocol)
+	if sockPath == "" {
+		return fmt.Errorf("cannot get sockPath of protocol %s", protocol)
+	}
+
+	dc, ctx, conn, cancelFunc, err := dmiclient.GenerateDMIClient(sockPath)
+	if err != nil {
+		return err
+	}
+
+	defer conn.Close()
+	defer cancelFunc()
+
+	udr, err := dmiclient.UpdateDeviceRequest(device)
+	if err != nil {
+		return fmt.Errorf("fail to generate UpdateDeviceRequest for device %s with err: %v", device.Name, err)
+	}
+	_, err = dc.UpdateDevice(ctx, udr)
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+func GrpcCreateDeviceModel(model *v1alpha2.DeviceModel) error {
+	protocol := model.Spec.Protocol
+	sockPath := GetSockPath(protocol)
+	if sockPath == "" {
+		return fmt.Errorf("cannot get sockPath of protocol %s", protocol)
+	}
+
+	dc, ctx, conn, cancelFunc, err := dmiclient.GenerateDMIClient(sockPath)
+	if err != nil {
+		return err
+	}
+
+	defer conn.Close()
+	defer cancelFunc()
+
+	cdmr, err := dmiclient.CreateDeviceModelRequest(model)
+	if err != nil {
+		return fmt.Errorf("fail to create CreateDeviceModelRequest for device model %s with err: %v", model.Name, err)
+	}
+	_, err = dc.CreateDeviceModel(ctx, cdmr)
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+func GrpcRemoveDeviceModel(model *v1alpha2.DeviceModel) error {
+	protocol := model.Spec.Protocol
+	sockPath := GetSockPath(protocol)
+	if sockPath == "" {
+		return fmt.Errorf("cannot get sockPath of protocol %s", protocol)
+	}
+
+	dc, ctx, conn, cancelFunc, err := dmiclient.GenerateDMIClient(sockPath)
+	if err != nil {
+		return err
+	}
+
+	defer conn.Close()
+	defer cancelFunc()
+
+	rdmr, err := dmiclient.RemoveDeviceModelRequest(model.Name)
+	if err != nil {
+		return fmt.Errorf("fail to create RemoveDeviceModelRequest for device model %s with err: %v", model.Name, err)
+	}
+	_, err = dc.RemoveDeviceModel(ctx, rdmr)
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+func GrpcUpdateDeviceModel(model *v1alpha2.DeviceModel) error {
+	protocol := model.Spec.Protocol
+	sockPath := GetSockPath(protocol)
+	if sockPath == "" {
+		return fmt.Errorf("cannot get sockPath of protocol %s", protocol)
+	}
+
```

Comment:
use GetDMIClientByProtocol to get client

Reply:
done